keps/sig-network/3015-node-level-topology/README.md

```diff
@@ -0,0 +1,436 @@
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests for meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The `Local` policy for `service.spec.externalTrafficPolicy` and
+`service.spec.internalTrafficPolicy` allows configuring services so
+that they will always deliver to local endpoints, avoiding an
+unnecessary hop to another node when the request could just as easily
+be processed locally.
+
+However, they both fail entirely if there are no endpoints for the
+service on the node. This tradeoff is unacceptable for some use cases.
+
+The [Topology Aware Hints feature] provides a way of signaling to
+kube-proxy that it should optimize service delivery on a
+per-availability zone basis, but does not provide any way to optimize
+on any other basis.
+
+This KEP adds a new topology hint, to tell kube-proxy that a Service
+is expected to have an endpoint on every node most of the time, and so
+it should optimize traffic by always routing to a local endpoint _when
+one is available_, but using other topology hints or ordinary
+`Cluster` traffic policy when not.
+
+[Topology Aware Hints feature]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2433-topology-aware-hints
+
+## Motivation
+
+### Goals
+
+- Allow configuring a service so that connections will be delivered to
+  a local endpoint when possible, and a remote endpoint if not.
+
+### Non-Goals
+
+N/A
```

Comment:
> This is essentially just an extension of ExternalTrafficPolicy=Local

Well, no. The discussion in #3016 seemed to me to be saying that local traffic policy implies a _semantic_ distinction between local and remote delivery, and the fact that it is more efficient is really just a side effect. Whereas this is solely about efficiency.

Reply:
As currently written, this can't serve as a replacement for `externalTrafficPolicy: Local`, because it doesn't coordinate with the LoadBalancer parts, so you'd have no way to force incoming connections to end up on the right node.

We could adjust things so LoadBalancer services do try to take advantage of node-level topology, but if we were going to do that, it seems like really we should just go back to `externalTrafficPolicy: PreferLocal`.

I guess you could make the argument that for _external_ traffic policy, this _is_ a semantic thing, not just a topology thing; you don't just want to change which endpoints get picked, you also want to change how the cloud load balancers work.

OTOH, for internal traffic, the desired behavior for the primary use case (DNS) really is just "better topology". OTOOH, I am still totally unconvinced that there is any _real_ use case for `internalTrafficPolicy: Local` as opposed to node-level topology.

So maybe the real answer is:
1. add `externalTrafficPolicy: PreferLocal`
2. add node-level topology for services like DNS
3. drop `internalTrafficPolicy`

Reply:
(although also, it was argued that ProxyTerminatingEndpoints mostly gets rid of the need for `externalTrafficPolicy: PreferLocal` anyway, by fixing the awkward bits of `externalTrafficPolicy: Local` so you _can_ use it just as an optimization if you want.)

Reply:
> As currently written, this can't serve as a replacement for `externalTrafficPolicy: Local`

(...which is obviously a bug since one of the User Stories talks about doing that)